File: chapter5/InterestingParty.py
"""
 SRM494 Div 2 Level 1
"""

import time

def best_invitation(first, second):
    answer = 0
    for i in xrange(len(first)):
        f = 0
        s = 0
        for j in xrange(len(first)):
            if first[i] == first[j]:
                f += 1
            if first[i] == second[j]:
                f += 1
            if second[i] == first[j]:
                s += 1
            if second[i] == second[j]:
                s += 1

        answer = max(f, s, answer)
        return answer

if __name__ == '__main__':
    count = 0
    for line in open('../testcase/InterestingParty.txt', 'r'):
        count += 1
        if count == 1:
            first = line[: -1].split(' ')
        if count == 2:
            second = line[: -1].split(' ')
        if count == 3:
            answer = int(line)

            start = time.time()
            result = best_invitation(first, second)
            end = time.time()

            if result == answer:
                print 'PASS!'
            else:
                print 'FAILURE!'

            print ("Time: {0}".format((end - start) * 1000)) + "[ms]"
            count = 0

File: chapter5/InterestingParty.java
/*
 * SRM494 Div 2 Level 1
 */

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;

public class InterestingParty {
  public static int bestInvitation(String[] first, String[] second) {
    int answer = 0;
    for (int i = 0; i < first.length; i++) {
      int f = 0;
      int s = 0;
      for (int j = 0; j < first.length; j++) {
        if (first[i].equals(first[j])) f++;
        if (first[i].equals(second[j])) f++;
        if (second[i].equals(first[j])) s++;
        if (second[i].equals(second[j])) s++;
      }
      answer = Math.max(f, answer);
      answer = Math.max(s, answer);
    }
    return answer;
  }

  public static void main(String[] args) {
    try(BufferedReader br = Files.newBufferedReader(
            Paths.get("../testcase/InterestingParty.txt"),
            StandardCharsets.UTF_8)) {
      String line;
      String[] first = null;
      String[] second = null;
      int answer;

      int count = 0;
      while ((line = br.readLine()) != null) {
        count++;
        if (count == 0) {
          first = line.split(" ");
        } else if (count == 1) {
          second = line.split(" ");
        } else {
          answer = Integer.parseInt(line);
          count = 0;

          long start = System.nanoTime();

          int result = bestInvitation(first, second);

          long end = System.nanoTime();

          if (answer == result) {
            System.out.println("PASS!!");
          } else {
            System.out.println("FAILURE!!");
          }

          System.out.println("Time: " + (end - start) / 1000000f + "[ms]");
        }
      }
    } catch(IOException e) {
      e.printStackTrace();
    }
  }
}
